doc/pekwm_screenshot.1
.nh
.TH pekwm\_sreenshot 1 "February 2021" pekwm\_sreenshot "User Manual"

.SH NAME
.PP
pekwm\_screenshot \- a simple screenshot application


.SH SYNOPSIS
.PP
pekwm\_screenshot [\-\-display] [output.png]


.SH DESCRIPTION
.PP
pekwm\_screenshot is a simple screenshot application bundled together
with pekwm.

.PP
Run without any arguments pekwm\_screenshot will output an image in the
current directory named pekwm\_screenshot\-YYMMDDTHHMMSS\-WIDTHxHEIGHT.png
where WIDTH and HEIGHT correspond to the size of the display.

.PP
If a file name is given it will be used as is without any processing
supported.


.SH OPTIONS
.PP
\fB\-\-help\fP Show help information.

.PP
\fB\-\-display\fP \fIDISPLAY\fP Connect to DISPLAY instead of DISPLAY set in environment.

.PP
\fB\-\-wait\fP \fIseconds\fP to wait before taking screenshot.


NEWS
NEWS pekwm-0.1.7
================

Bugs fixed

  * #XX Several memory leaks identified with valgrind plugged.
  * #XX Fixed broken iconfied state reported by
     <giuliani{.}blackfenix{@}gmail{.}com>
  * #XX CfgParser parsed data twice.
  * #XX StatusWindow placement fixed.
  * #XX Fix for _NET_WM_NAME ops by Florian Hoppe <flp{@}floppi{.}de>
  * #22 Font type is now possible to define as the first argument in
     themes.
  * #30 Panel raise issues, reported and fixed by Jyri Jokinen
     <shared{@}adresh{.}com>
  * #21 Dynamic menu entries become unselectable after first
     visit. Reported by Lassi Pölönen and fixed by Andreas
     <ioerror{@}lavabit{.}com>
  * #14 SendKey action requested by nidan.
  * #29 drag'n'drop requires the recieving window to be moved before
     working. Reported by Jyri Jokinen <shared{@}adresh{.}com>, fixed
     by Andreas <ioerror{@}lavabit{.}com>
  * #32 Shaping issues after shading. Reported by Paulo Igor Barra
     Nascimento <pibarnas{@}yahoo{.}com{.}br> and fixed by Andreas
     <ioerror{@}lavabit{.}com>
  * #XX Corrects the interpretation of the motif windowmanager hints
     by Andreas <ioerror{@}lavabit{.}com>
  * #XX Remove perl scripts reducing dependencies on packages. 

User visible changes

  * UTF-8 support added for reading and rendering window titles.
  * _NET_WM_ICON is now read and rendered in menus.
  * RandR support for multi-screen configurations.
  * New default theme.
  * CmdDialog size is now calculated from the current head and not the
    whole display.

Configuration changes

  * _NET_WM_WINDOW_TYPE handling configurable via autoproperties.
  * Width/Height 0 in autoproperties now means screen width/height.
  * WorkspacesPerRow and WorkspaceNames added with matching
    GotoWorkspace/SendToWorkspace parameters Up, PrevV, Down, NextV.
  * ShowWorkspaceIndicator
  * WorkspaceIndicatorScale
  * PCRE support remove.
  * EdgeIndent option added to Screen section of main configuration
    file. If true reserves space where the screen edge is.
  * EdgeSize now supports per edge configuration. 

Theme changes

  * Shaping suport for non-corner windows.
  * XMB type is used by default in favour of X11 type fonts.
  * Font alpha is now handles as % from 0-100 instead of previous Xft
    unsigned int value.
  * HeightAdapt option added to theme.
  * Font type is now possible to define as the first argument in themes.


src/screenshot/pekwm_screenshot.cc
//
// pekwm_screenshot.cc for pekwm
// Copyright (C) 2021-2025 Claes Nästén <pekdon@gmail.com>
//
// This program is licensed under the GNU GPL.
// See the LICENSE file for more information.
//

#include "Compat.hh"
#include "Util.hh"
#include "X11.hh"

#include "../tk/ImageHandler.hh"
#include "../tk/PImageLoaderPng.hh"

#include <iostream>
#include <iomanip>
#include <sstream>

extern "C" {
#include <sys/types.h>
#include <getopt.h>
#include <time.h>
#include <unistd.h>
#include <X11/Xutil.h>
}

static ImageHandler* _image_handler = nullptr;

namespace pekwm
{
	ImageHandler* imageHandler()
	{
		return _image_handler;
	}
}

static void init(Display* dpy)
{
	_image_handler = new ImageHandler();
}

static void cleanup()
{
	delete _image_handler;
}

static void usage(const char* name, int ret)
{
	std::cout << "usage: " << name << " [-dhw] [screenshot.png]"
		  << std::endl;
	std::cout << "  -d --display dpy    Display" << std::endl;
	std::cout << "  -h --help           Display this information"
		  << std::endl;
	std::cout << "  -w --wait seconds   Wait seconds before taking "
		     "screenshot" << std::endl;
	exit(ret);
}

static std::string get_screenhot_name(const Geometry& gm)
{
	time_t t = time(nullptr);
	tm tm;
	localtime_r(&t, &tm);

	std::ostringstream name;
	name << "pekwm_screenshot-";
	name << std::put_time(&tm, "%Y%m%dT%H%M%S");
	name << "-" << gm.width << "x" << gm.height;
	name << ".png";
	return name.str();
}

static int take_screenshot(const std::string& output)
{
	Geometry gm = X11::getScreenGeometry();
	XImage *ximage = X11::getImage(X11::getRoot(),
				       gm.x, gm.y, gm.width, gm.height,
				       AllPlanes, ZPixmap);
	if (ximage == nullptr) {
		std::cerr << "Failed to take a screenshot" << std::endl;
		return 1;
	}

	PImage *image = new PImage(ximage);
	X11::destroyImage(ximage);

	bool success =
		PImageLoaderPng::save(output, image->getData(),
				      image->getWidth(), image->getHeight());
	return success ? 0 : 1;
}

int main(int argc, char* argv[])
{
	// Limit access, limit further after X11 connection is setup.
	pledge_x11_required("");

	const char* display = NULL;
	int wait_seconds = 0;

	static struct option opts[] = {
		{const_cast<char*>("display"), required_argument, nullptr,
		 'd'},
		{const_cast<char*>("help"), no_argument, nullptr, 'h'},
		{const_cast<char*>("wait"), required_argument, nullptr, 'w'},
		{nullptr, 0, nullptr, 0}
	};

	int ch;
	while ((ch = getopt_long(argc, argv, "d:hw:", opts, nullptr)) != -1) {
		switch (ch) {
		case 'd':
			display = optarg;
			break;
		case 'h':
			usage(argv[0], 0);
			break;
		case 'w':
			try {
				wait_seconds = std::stoi(optarg);
			} catch (std::invalid_argument&) {
				usage(argv[0], 1);
			}
			break;
		default:
			usage(argv[0], 1);
			break;
		}
	}

	if (! X11::init(display, std::cerr)) {
		return 1;
	}

	// X11 connection has been setup, limit access further
	pledge_x("stdio rpath wpath cpath", "");

	init(X11::getDpy());

	std::string output;
	if (optind < argc) {
		output = argv[optind];
	} else {
		output = get_screenhot_name(X11::getScreenGeometry());
	}

	if (wait_seconds > 0) {
		if (wait_seconds > 3) {
			sleep(wait_seconds - 3);
			wait_seconds = 3;
		}
		for (int i = wait_seconds; i > 0; i--) {
			std::cout << i << "..." << std::flush;
			sleep(1);
		}
		std::cout << std::endl;
	}

	int ret = take_screenshot(output);
	if (ret) {
		std::cerr << "failed to write screenshot to " << output
			  << std::endl;
	} else {
		std::cout << "screenshot written to " << output << std::endl;
	}

	cleanup();
	X11::destruct();

	return ret;
}
